client/src/main/java/io/confluent/kafka/schemaregistry/client/security/bearerauth/oauth/OauthTokenCache.java

```diff
@@ -0,0 +1,122 @@
+ * setCurrentToken(OAuthBearerToken)} to set a newly retrieved token. {@link
+ * #calculateTokenExpiryTime calculateTokenRenRefreshTime()} is calculated from the token's  at time
+ * of setting the token using {@link #setCurrentToken(OAuthBearerToken)
+ * setCurrentToken(OAuthBearerToken)}. In order to encourage fetching a new token in advance before
+ * the token expiry, {@link #calculateTokenExpiryTime calculateTokenRenRefreshTime()} is altered by
+ * multiplying {@link #CACHE_EXPIRY_THRESHOLD CACHE_EXPIRY_THRESHOLD()}. Thus we consider a shorter
+ * lifespan than actual.
+ */
+
+public class OauthTokenCache {
+
+  public static final float CACHE_EXPIRY_THRESHOLD = 0.8f;
+  private static final Logger log = LoggerFactory.getLogger(OauthTokenCache.class);
+  private final short cacheExpiryBufferSeconds;
+  private OAuthBearerToken currentToken;
+  private long cacheExpiryMs;
+
+  public OauthTokenCache(short cacheExpiryBufferSeconds) {
+    /*
+     setting a value of 0 so that isTokenExpired() returns true in when called the first time
+     cacheExpiryMs time of 0L means this cache does not have a valid token.
+    */
+    this.cacheExpiryMs = 0L;
+    this.cacheExpiryBufferSeconds = cacheExpiryBufferSeconds;
+  }
+
+  public OAuthBearerToken getCurrentToken() {
+    return currentToken;
+  }
+
+  public void setCurrentToken(OAuthBearerToken currentToken) {
+    if (currentToken == null) {
+      // if token null, set cacheExpiryMs to initial state.
+      cacheExpiryMs = 0L;
+      return;
+    }
+
+    this.cacheExpiryMs = calculateTokenExpiryTime(currentToken);
+    this.currentToken = currentToken;
+  }
+
+  //visible for testing
+  protected long calculateTokenExpiryTime(OAuthBearerToken currentToken) {
+    long nowMs = Instant.now().toEpochMilli();
+    long tokenExpiryMs = currentToken.lifetimeMs();
+    // If the current time is greater than tokenExpiryMs we log the possible clock skew
+    // message return the maximum time i.e tokenExpiryMs
+    if (nowMs > tokenExpiryMs) {
+      log.error(
+          "Schema Registry OAuth Token [Principal={}]: Current clock: {} is later than expiry {}. "
+              + "This may indicate a clock skew problem."
+              + " Check that this host's and remote host's clocks are in sync."
+              + " This process is likely unable to authenticate SASL connections (for example, it "
+              + "is unlikely to be able to authenticate a connection with a Schema Registry).",
+          currentToken.principalName(), new Date(nowMs), new Date(tokenExpiryMs));
+
+      return tokenExpiryMs;
+    }
+
+    Long optionalStartTime = currentToken.startTimeMs();
+    long startMs = optionalStartTime != null ? optionalStartTime.longValue() : nowMs;
+
+    long cacheExpiryMs;
+    if (nowMs + 1000L * cacheExpiryBufferSeconds > tokenExpiryMs) {
+      cacheExpiryMs = nowMs + (long) Math.floor((tokenExpiryMs - nowMs) * CACHE_EXPIRY_THRESHOLD);
```

Comment:
Here after using an absolute time wall between token expiry and current time, we are calculating the time remaining to expire and multiplying it by CACHE_EXPIRY_THRESHOLD.
Now, this could be different each time and it will make tuning CACHE_EXPIRY_THRESHOLD extremely difficult. 
As if the time remaining is let's say 5 mins and CACHE_EXPIRY_THRESHOLD is 80% then we will expire 1 min before. 
But if time remaining is 2 mins we will expire 24 seconds before. So each time it will be a different thing. So instead of this we can go with a simple approach and return nowMs here. So that we have more control over when the cache should expire.